tensorflow/compiler/mlir/xla/ir/infer_fusibility_op_interface.td

```diff
@@ -0,0 +1,171 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+// This file contains inferFusiblityOpInterface, which is used to guide
+// fusion decsion.
+
+#ifndef MLIR_INFER_FUSIBILITY_OP_INTERFACE
+#define MLIR_INFER_FUSIBILITY_OP_INTERFACE
+
+include "mlir/IR/OpBase.td"
+
+// OpInterface to query if an op is fusible and to query the shape equality
+// constraint among the inputs and outputs of an op.
+def InferFusibilityOpInterface : OpInterface<"InferFusibilityOpInterface"> {
+  let description = [{
+    Interface to query if an op is fusible and to query the shape equality
+    constraint among the inputs and outputs of an op.
+  }];
+
+  let methods = [
+    InterfaceMethod<
+      /*desc=*/[{If true, this op can be fused with its operands
+      }],
+      /*retTy=*/"bool",
+      /*methodName=*/"isFusibleWithOperand",
+      /*args=*/(ins),
+      /*methodBody=*/[{}],
+      /*defaultImplementation=*/[{
+        /// Returns whether this op can be fused with its operands
+        return true;
+      }]
+    >,
+    InterfaceMethod<
+      /*desc=*/[{If true, this op can be fused with its consumers
+      }],
+      /*retTy=*/"bool",
+      /*methodName=*/"isFusibleWithConsumer",
+      /*args=*/(ins),
+      /*methodBody=*/[{}],
+      /*defaultImplementation=*/[{
+        /// Return whether this op can be fused withh its consumers
+        return true;
+      }]
+    >,
+    InterfaceMethod<
+      /*desc=*/"Return whether two inputs have the same shape (assuming no"
+               "implicit broadcasting).",
+      /*retTy=*/"bool",
+      /*methodName=*/"inferInputsShapeEquality",
+      /*args=*/(ins "int":$lhs, "int":$rhs),
+      /*methodBody=*/[{}],
+      /*defaultImplementation=*/[{
+        /// Return whether two inputs have the same shape.
+        Operation *op = this->getOperation();
+        assert(lhs < op->getNumOperands() && lhs >= 0 &&
+               rhs < op->getNumOperands() && rhs >= 0);
+        if (lhs == rhs) return true;
+
+        // if both lhs and rhs have static shapes, check them directly
+        Type lhs_ty = op->getOperand(lhs).getType();
+        Type rhs_ty = op->getOperand(rhs).getType();
+        auto lhs_shape_type = lhs_ty.dyn_cast_or_null<RankedTensorType>();
+        auto rhs_shape_type = rhs_ty.dyn_cast_or_null<RankedTensorType>();
+        if (!lhs_shape_type || !lhs_shape_type.hasStaticShape() ||
+            !rhs_shape_type || !rhs_shape_type.hasStaticShape() ||
+            lhs_shape_type.getRank() != rhs_shape_type.getRank()) {
+          return false;
+        }
+        for (int i = 0, e = lhs_shape_type.getRank(); i < e; ++i)
```

Comment:
Why not just return lhs_shape_type.getShape() == rhs_shape_type.getShape() ? (here and below)